Add automated tests for the entire volume workflow

Ebaneck opened the pull request:
**Component**:

'test', 'storage', 'kubernetes'
<!-- E.g. 'salt', 'containers', 'kubernetes', 'build', 'tests'... -->

**Context**: 

See issue #1417 
- The ticket contains an extensive description of some of the test scenarios we would like to have.

**Summary**:

- We would love to automate CI tests for volume creation, volume deletion, persistent volume creation and deletion including test cases where semantic errors might occur like a bad storage class naming, invalid filesystem types which we do not support and much more.

How to test?
`tox -e tests -- -m 'volume'
`

**Acceptance criteria**: 

- CI tests should continue to work normally
- All volume workflow scenarios should be accounted for in the test cases described. 

---

<!-- Declare one or more issues to close once this PR gets merged -->

Closes: #1417

Ebaneck committed:
test/volume: test volume creation

Refs: #1417

bert-e commented:
# Hello ebaneck,



My role is to assist you with the merge of this
pull request. Please type `@bert-e help` to get information
on this process, or consult the [user documentation](
https://eve.devsca.com/github/scality/metalk8s/bert-e/doc/user).





*Status report is not available.*

bert-e commented:
# Waiting for approval



The following approvals are needed before I can proceed with the merge:


* the author


* one peer





Peer approvals *must* include at least 1 approval from the following list:

* @teddyandrieux

* @monpote

* @jbertran

* @sayf-eddine-scality

* @nicolast

* @cmonfort

* @slaperche-scality

* @thomasdanan

* @zempashi

* @lucieleonard

* @gdemonet

* @carlito-scality

* @nootal

* @alexandre-allard-scality

Ebaneck commented:
@scality/metalk8s and any other person reviewing this,  feel free to leave a comment stating if this is the right path to follow and/or some test scenarios we have not accounted for as specified in the original ticket.

slaperche-scality commented on tests/post/features/volume.feature:
```diff
@@ -11,15 +11,6 @@ Feature: Volume management
-              storageClassName: metalk8s-prometheus
-              sparseLoopDevice:
-                size: 10Gi
+        When I create a Volume 'volume1'
```
No, when you create a volume you want to specify the body.

This should be left untouched

slaperche-scality replied:
This is not resolved: you didn't restore the step as it was…

slaperche-scality commented on tests/post/steps/test_volume.py:
```diff
@@ -105,4 +100,38 @@ def _check_size():
+# Helpers {{{
+
+
+def _create_volume(host, name, k8s_custom_client):
```
not exactly, here we want an helper to create a volume from a body.

```python
def _create_volume(k8s_client, body):
    try:
        k8s_client.create_cluster_custom_object(
            group="storage.metalk8s.scality.com",
            version="v1alpha1",
            plural="volumes",
            body=body)
        )
    except (ApiException, HTTPError) as exc:
        if isinstance(exc, ApiException) and exc.status == 404:
            assert False, 'Exception while creating volume {}'.format(name)
        raise
```

slaperche-scality commented on tests/post/steps/test_volume.py:
```diff
@@ -105,4 +100,38 @@ def _check_size():
+# Constants {{{
+
+
+DEFAULT_VOLUME = """
```
This should be introduced in the next commit, because you don't need it here; this commit only introduce the helper.

Also, constants are usually defined at the top of the fileso I would move the `Constants` fold before the `Fixtures` one.

slaperche-scality commented on tests/post/steps/test_volume.py:
```diff
@@ -34,14 +34,9 @@ def test_volume_creation(host):
-    )
+@when(parsers.parse("I create a Volume '{name}'"))
+def create_volume(host, name, k8s_custom_client):
+    _create_volume(host, name, k8s_custom_client)
```
This should become `_create_volume(k8s_custom_client, body)`

slaperche-scality commented on tests/post/features/volume.feature:
```diff
@@ -11,15 +11,6 @@ Feature: Volume management
-              storageClassName: metalk8s-prometheus
-              sparseLoopDevice:
-                size: 10Gi
+        When I create a Volume 'volume1'
```
You should explain why we suddenly introduce an helper.

Something like:

```
test/volume: refactor volume creation test to use helpers 
In future test we will need to create Volume by giving only a name
(we will use a default template for the other fields) as well as
creating Volume from a YAML manifest (what we are already doing).

Since the creation code will be the same, let's factor in into an helper.
```

slaperche-scality commented on tests/post/steps/test_volume.py:
```diff
@@ -100,6 +110,31 @@ def _check_size():
 
+
+@then(parsers.parse("the Volume '{name}' does not exists"))
+def volume_not_exist(host, name, k8s_custom_client, k8s_client):
```
`check_volume_absent`, `then` statement usually have function name starting with `check`.

slaperche-scality commented on tests/post/steps/test_volume.py:
```diff
@@ -134,4 +169,38 @@ def _create_volume(host, name, k8s_custom_client):
         raise
+
+
+def _volume_exist(host, name, k8s_custom_client):
```
This return an object, not a Boolean: call it `_get_volume`.

`host` is unused: can be removed.
Here we're not using the fixture (it's not a `then`/`when`/`given`, so fixture are not magically applied): we can name the parameter `k8s_client`).


Also we know that we're fetching a Volume here, so the variable can be named `volume` instead of `custom_object`

```python
 def _get_volume(k8s_client, name):
    try:
        return k8s_custom_client.get_cluster_custom_object(
            group="storage.metalk8s.scality.com",
            version="v1alpha1",
            plural="volumes",
            name=name
        )
    except (ApiException, HTTPError) as exc:
        if isinstance(exc, ApiException) and exc.status == 404:
            return None
        raise
```

Ebaneck replied:
> This return an object, not a Boolean: call it `_get_volume`.
> 
> `host` is unused: can be removed.
> Here we're not using the fixture (it's not a `then`/`when`/`given`, so fixture are not magically applied): we can name the parameter `k8s_client`).

Not sure this works as stated, tested and it only works with the right fixtures (`k8s_custom_client`). Safe to say that for some reason, fixtures are magically applied here.

> 
> Also we know that we're fetching a Volume here, so the variable can be named `volume` instead of `custom_object`
> 
> ```python
>  def _get_volume(k8s_client, name):
>     try:
>         return k8s_custom_client.get_cluster_custom_object(
>             group="storage.metalk8s.scality.com",
>             version="v1alpha1",
>             plural="volumes",
>             name=name
>         )
>     except (ApiException, HTTPError) as exc:
>         if isinstance(exc, ApiException) and exc.status == 404:
>             return None
>         raise
> ```

slaperche-scality replied:
I applied this diff (which is what I was recommending) on your current PR:

```diff
--- a/tests/post/steps/test_volume.py
+++ b/tests/post/steps/test_volume.py
@@ -160,9 +160,9 @@ def check_pv_absent(name, k8s_client):
 # Helpers {{{
 
 
-def _create_volume(body, k8s_custom_client):
+def _create_volume(body, k8s_client):
     try:
-        k8s_custom_client.create_cluster_custom_object(
+        k8s_client.create_cluster_custom_object(
             group="storage.metalk8s.scality.com",
             version="v1alpha1",
             plural="volumes",
@@ -174,9 +174,9 @@ def _create_volume(body, k8s_custom_client):
         raise
 
 
-def _get_volume(name, k8s_custom_client):
+def _get_volume(name, k8s_client):
     try:
-        return k8s_custom_client.get_cluster_custom_object(
+        return k8s_client.get_cluster_custom_object(
             group="storage.metalk8s.scality.com",
             version="v1alpha1",
             plural="volumes",
@@ -187,9 +187,8 @@ def _get_volume(name, k8s_custom_client):
             return None
 
 
-def _delete_volume(name, k8s_custom_client):
-
-    k8s_custom_client.delete_cluster_custom_object(
+def _delete_volume(name, k8s_client):
+    k8s_client.delete_cluster_custom_object(
         group="storage.metalk8s.scality.com",
         version="v1alpha1",
         plural="volumes",
```

And it works…

So no, fixtures are **not** magically applied :slightly_smiling_face:

slaperche-scality commented on tests/post/steps/test_volume.py:
```diff
@@ -31,13 +32,32 @@ def test_deploy_operator(host):
+
+@when(parsers.parse("I delete the Volume '{name}'"))
+def delete_volume(host, name, k8s_custom_client):
+    _delete_volume(host, name, k8s_custom_client)
```
`_delete_volume` is only used here, so I don't think you need an helper.

Simply write:

```python
@when(parsers.parse("I delete the Volume '{name}'"))
def delete_volume(host, name, k8s_custom_client):
        k8s_custom_client.delete_cluster_custom_object(
            group="storage.metalk8s.scality.com",
            version="v1alpha1",
            plural="volumes",
            name=name,
            body=kubernetes.client.V1DeleteOptions(),
            grace_period_seconds=0
        )
```

We don't need to handle exception here, if something fail the test should fail.

Ebaneck replied:
yes we do not need the helper yet , but the next test case makes use of a delete function definition hence the helper

slaperche-scality replied:
> the next test case makes use of a delete function definition hence the helper

Then introduce it in the next test case.
I'm not even sure the next test case use it actually.

The next test case use ` When I delete the Volume 'volume3'` which is exactly what we use here. No code to add for that.

slaperche-scality commented on tests/post/steps/test_volume.py:
```diff
@@ -31,13 +32,32 @@ def test_deploy_operator(host):
+def volume_exist(host, name, k8s_custom_client):
+    volume = _volume_exist(host, name, k8s_custom_client)
+    if volume is None:
+        _create_volume(host, name, k8s_custom_client)
```
Here you should have:

```python
if volume is None:
    body = DEFAULT_VOLUME.format(name=name)
    _create_volume(k8s_custom_client, body)
```

slaperche-scality replied:
We also probably need to have a wait here, until the Volume becomes `Available`.

Ebaneck replied:
I fail to see why we need to wait here. We are making a request to the API and once the request is submitted, we should terminate the transaction.

Checking the Availability of the created volume should be handled in the next stage which is what we do actually.

slaperche-scality replied:
Nope, the role of the `Given` is to ensure that the volume exists and thus should wait for it to becomes available.

The test scenario is:

```cucumber
Scenario: Test volume deletion (sparseLoopDevice)
    Given a Volume 'volume2' exists
    When I delete the Volume 'volume2'
    Then the Volume 'volume2' does not exists
    And the PersistentVolume 'volume2' does not exists
```

We try to delete right after we are sure that the Volume exists: there is no explicit `wait` to add in the scenario because it must be handled by the given.

slaperche-scality commented on tests/post/steps/test_volume.py:
```diff
@@ -100,6 +110,31 @@ def _check_size():
+            assert False, 'Volume {} still exist'.format(name)
+    utils.retry(
+        _check_volume, times=30, wait=2,
+        name='checking if volume {} exist'.format(name)
```
`checking the absence of volume {}`

slaperche-scality commented on tests/post/steps/test_volume.py:
```diff
@@ -100,6 +110,31 @@ def _check_size():
+
+
+@then(parsers.parse("the PersistentVolume '{name}' does not exists"))
+def check_pv(host, name, k8s_client):
```
Hum, this should be similar to `check_volume_absent`, but for the PV instead (so rename it, use a `retry`, …)

bert-e commented:
# Waiting for approval



The following approvals are needed before I can proceed with the merge:


* the author


* one peer





Peer approvals *must* include at least 1 approval from the following list:

* @teddyandrieux

* @monpote

* @jbertran

* @sayf-eddine-scality

* @nicolast

* @cmonfort

* @slaperche-scality

* @thomasdanan

* @zempashi

* @lucieleonard

* @gdemonet

* @carlito-scality

* @nootal

* @alexandre-allard-scality





The following reviewers are expecting changes from the author, or must review again:

* @slaperche-scality

Ebaneck committed:
test/volume: test volume deletion

We move the volume retrieval and creation code into helpers in order to
reuse them without code duplication.

Refs: #1417

Ebaneck committed:
test/volume: test PV protection

Test that a PV created by a Volume cannot be deleted manually (I mean,
you can delete it, but it'll only be marked for deletion and be retained
until the Volume is deleted).

We move the PV retrieval into an helper for the sake of reusability.

Refs: #1417

Ebaneck committed:
test/volume: Create a volume without volume type

Refs: #1417

slaperche-scality committed:
test: relax Gherkin language validation globally

Because having implicit dependencies between test scenarios is pretty
bad (trust me, been there done that…) we want to make test scenarios
self-contained (as much as possble).

And to do so, we sometimes need to have more than one sequence of
`When/Then` in a single scenario, but `pytest.bdd` (unlike `behave`…)
forbids that by default.

So let's relax the validation to allow this kind of pattern.

Refs: #1417
Signed-off-by: Sylvain Laperche <sylvain.laperche@scality.com>

Ebaneck committed:
test/volume: create a volume with an invalid type

Refs: #1417

charlesprost commented on tests/post/features/volume.feature:
```diff
@@ -8,3 +8,59 @@ Feature: Volume management
+    Scenario: Test volume deletion (sparseLoopDevice)
+        Given a Volume 'volume2' exists
+        When I delete the Volume 'volume2'
+        Then the Volume 'volume2' does not exists
```
```suggestion
        Then the Volume 'volume2' does not exist
```

charlesprost commented on tests/post/features/volume.feature:
```diff
@@ -8,3 +8,59 @@ Feature: Volume management
+        Given a Volume 'volume2' exists
+        When I delete the Volume 'volume2'
+        Then the Volume 'volume2' does not exists
+        And the PersistentVolume 'volume2' does not exists
```
```suggestion
        And the PersistentVolume 'volume2' does not exist
```

charlesprost replied:
Same below

charlesprost commented on tests/post/steps/test_volume.py:
```diff
@@ -1,10 +1,39 @@
+
+# Constants {{{
+
+DEFAULT_VOLUME = """
```
If there is a default volume we might want to avoid passing it in the scenarios themselves.

slaperche-scality replied:
Nope, two different use cases here.

You want to create a volume, and that's the point of the test?
=> then pass a YAML

Using a templated stuff for that will just give unreadable long line like:
`And I create a volume titled 'test-volume' of type 'sparseLoopDevice' with storageclass 'random-storageclass' of size '1Gi' for the 'bootstrap' node with a random device type 'rawSparseDevice'`
So, please no…

On the other hand, when you just need a volume to exists (because you want delete it or whatever) then we can use an hardcoded default one.

charlesprost commented on tests/post/steps/test_volume.py:
```diff
@@ -15,7 +44,86 @@ def test_deploy_storage_class(host):
+@given(parsers.parse("a Volume '{name}' exists"))
+def volume_exist(host, name, k8s_custom_client):
+    volume = _get_volume(k8s_custom_client, name)
+    if volume is None:
```
Python tip: when 90% of the function is indented after a `if`, it could be better to do something like:

```
if volume is not None:
    return
#blablabla
```

charlesprost commented on tests/post/steps/test_volume.py:
```diff
@@ -26,3 +134,152 @@ def check_storage_class(host, name, k8s_apiclient):
+
+@then(parsers.parse("the Volume '{name}' is '{status}'"))
+def check_volume_status(host, name, status, k8s_custom_client):
+    def _check_status():
```
Define only once `_check_status`

Ebaneck replied:
The scope of the function is local so, I think this should not be a problem.
 By the way, the check status function is specific in each case it has been used. Maybe we should really give this nested functions proper names

slaperche-scality replied:
>  Maybe we should really give this nested functions proper names

We should (and we started to do so with `_check_status` and `_check_size`), it's easier for grepping & jumping in the code.

charlesprost commented on tests/post/steps/test_volume.py:
```diff
@@ -26,3 +134,152 @@ def check_storage_class(host, name, k8s_apiclient):
+    )
+
+
+@then(parsers.parse("the Volume '{name}' does not exists"))
```
```suggestion
@then(parsers.parse("the Volume '{name}' does not exist"))
```

Ebaneck committed:
test/volume: test in-use Volume protection

A volume cannot be deleted while the underlying PV is used by a Pod.
If you try to delete such Volume, it will stay available until the PV is
no longer used (e.g.: the pod is terminated), at that point everything
will be cleaned up.

Refs: #1417

slaperche-scality commented on tests/post/steps/test_volume.py:
```diff
@@ -26,3 +56,54 @@ def check_storage_class(host, name, k8s_apiclient):
+            )
+
+    utils.retry(
+        _check_size, times=30, wait=2,
```
Seeing that PV doesn't trigger long action at creation time, I think we may reduce the retry parameters.

Maybe `times=10, wait=2` is enough.

slaperche-scality commented on tests/post/steps/test_volume.py:
```diff
@@ -12,7 +13,6 @@
 
 # Constants {{{
 
-
```
Why this change?
If you really want to to it, should be done in the previous commit.

slaperche-scality commented on tests/post/steps/test_volume.py:
```diff
@@ -48,13 +48,57 @@ def test_deploy_operator(host):
+    body = DEFAULT_VOLUME.format(name=name)
+    _create_volume(k8s_custom_client, body)
+
+    def _check_status(status='Available'):
```
I think it could be simplified to:

```python
    def _check_availability():
        volume = _get_volume(k8s_custom_client, name)
        assert volume['status']['phase'] == 'Available',\
            'Volume not available, status={}'.format(volume['status']['phase'])
```

slaperche-scality commented on tests/post/steps/test_volume.py:
```diff
@@ -48,13 +48,57 @@ def test_deploy_operator(host):
+
+    utils.retry(
+        _check_status, times=30, wait=2,
+        name='checking status of Volume {}'.format(name)
```
`checking availability of Volume {}`

slaperche-scality commented on tests/post/steps/test_volume.py:
```diff
@@ -117,6 +151,34 @@ def _check_size():
+        try:
+            pv = k8s_client.read_persistent_volume(name)
+        except (ApiException, HTTPError) as exc:
+            if isinstance(exc, ApiException) and exc.status == 404:
```
```python
if isinstance(exc, ApiException) and exc.status == 404:
    return None
raise
```

slaperche-scality commented on tests/post/steps/test_volume.py:
```diff
@@ -117,6 +151,34 @@ def _check_size():
+        if pv is not None:
+            assert False, 'PersistentVolume {} exist'.format(name)
+    utils.retry(
+        _check_pv, times=30, wait=2,
```
As before, I think we can decrease the retry parameters: `times=10, wait=1` or `times=10, wait=2`.

slaperche-scality commented on tests/post/steps/test_volume.py:
```diff
@@ -117,6 +151,34 @@ def _check_size():
+@then(parsers.parse("the Volume '{name}' does not exists"))
+def check_volume_absent(host, name, k8s_custom_client, k8s_client):
+    def _check_volume():
+        volume = _get_volume(k8s_custom_client, name)
```
Can be simplified to:

```python
    def _check_volume_absent():
        assert _get_volume(k8s_custom_client, name) is None,\
            'Volume {} still exist'.format(name)
```

slaperche-scality commented on tests/post/steps/test_volume.py:
```diff
@@ -130,7 +192,19 @@ def _create_volume(k8s_client, body):
+            plural="volumes",
+            name=name
+        )
+    except (ApiException, HTTPError) as exc:
```
You should re-raise if it's not a 404:

```python
    except (ApiException, HTTPError) as exc:
        if isinstance(exc, ApiException) and exc.status == 404:
            return None
        raise
```

slaperche-scality commented on tests/post/steps/test_volume.py:
```diff
@@ -107,3 +118,19 @@ def _check_size():
+        )
+    except (ApiException, HTTPError) as exc:
+        if isinstance(exc, ApiException) and exc.status == 404:
+            assert False, 'Exception while creating volume {}'.format(name)
```
While only error when 404? Other errors are also errors.

Should be defined as:

```python
def _create_volume(k8s_client, body):
    k8s_client.create_cluster_custom_object(
        group="storage.metalk8s.scality.com",
        version="v1alpha1",
        plural="volumes",
        body=yaml.safe_load(body)
    )
```

slaperche-scality commented on tests/post/steps/test_volume.py:
```diff
@@ -48,13 +48,57 @@ def test_deploy_operator(host):
+# Given {{{
+
+@given(parsers.parse("a Volume '{name}' exists"))
+def volume_exist(host, name, k8s_custom_client):
```
nested function are usually defined at the top of the outer function, not in the middle:

```python
def volume_exist(host, name, k8s_custom_client):
    def _check_availability():
        volume = _get_volume(k8s_custom_client, name)
        assert volume['status']['phase'] == 'Available',\
            'Volume not available, status={}'.format(volume['status']['phase'])

    volume = _get_volume(k8s_custom_client, name)
    if volume is not None:
        return
    body = DEFAULT_VOLUME.format(name=name)
    _create_volume(k8s_custom_client, body)
    utils.retry(
        _check_availability, times=30, wait=2,
        name='checking availability of Volume {}'.format(name)
    )
```

slaperche-scality commented on tests/post/steps/test_volume.py:
```diff
@@ -179,6 +195,31 @@ def _check_pv():
+def check_pv_deletion_timestamp(name, k8s_client):
+    def _check_deletion_timestamp():
+        try:
+            pv = k8s_client.read_persistent_volume(name)
```
Like you have a `_get_volume` helper, you should have a `_get_pv` helper.

```python
def _get_volume(k8s_client, name):
    try:
        return k8s_client.read_persistent_volume(name)
    except (ApiException, HTTPError) as exc:
        if isinstance(exc, ApiException) and exc.status == 404:
            return None
        raise
```

then you could write:
```python
    def _check_deletion_timestamp():
        pv = _get_pv(name)
        assert ps is not None, 'PersistentVolume {} not found'.format(name)
        assert pv.metadata.deletion_timestamp is not None,\
            'PersistentVolume {} is not marked for deletion'.format(name)
```

slaperche-scality commented on tests/post/steps/test_volume.py:
```diff
@@ -218,6 +223,33 @@ def _check_deletion_timestamp():
+@then(parsers.parse("the Volume '{name}' is 'Failed' "
+                    "with code '{code}' and message matches '{pattern}'"))
+def check_volume_error(host, name, code, pattern, k8s_custom_client):
+    def _check_status():
```
`def _check_error():`

slaperche-scality commented on tests/post/features/volume.feature:
```diff
@@ -115,3 +115,88 @@ Feature: Volume management
         Then the Volume 'volume8' does not exists
         And the PersistentVolume 'volume8' does not exists
+
+    Scenario: Complete e2e volume workflow
```
`E2E`, it's an acronym.
So: `Complete E2E volume workflow`.

slaperche-scality commented on tests/post/steps/test_volume.py:
```diff
@@ -94,6 +99,22 @@ def _check_status(status='Available'):
 
+
+
+@given(parsers.parse("I create the following object:\n{body}"))
```
no, don't use some generic stuff such as "object'
It would be good if the step could create arbitrary object, but that's not the case (yet).

Also a `given` is about a state, not about action thus you can't write `I create`.

Either:
- `@given(parsers.parse('The following PersistentVolumeClaim exist:\n{body}'))`
- `@then(parsers.parse('I create the following PersistentVolumeClaim:\n{body}'))`

slaperche-scality commented on tests/post/steps/test_volume.py:
```diff
@@ -94,6 +99,22 @@ def _check_status(status='Available'):
+    )
+
+
+@given(parsers.parse("I create the following Pod:\n{body}"))
```
Same remarks, either:
- `@given(parsers.parse("the following Pod exist:\n{body}"))`
- `@then(parsers.parse("I create the following Pod:\n{body}"))`

slaperche-scality commented on tests/post/steps/test_volume.py:
```diff
@@ -255,6 +285,24 @@ def _check_status():
+            )
+
+    utils.retry(
+        _check_deletion_timestamp, times=30, wait=2,
```
This should appear quite fast (we don't need to call Salt or whatever) so we can reduce the settings here: `times=10, wait=1` or `times=10, wait=2`

slaperche-scality commented on tests/post/steps/test_volume.py:
```diff
@@ -255,6 +285,24 @@ def _check_status():
+            assert False, 'Volume {} not found'.format(name)
+
+        assert volume['metadata']['deletionTimestamp'] is not None,\
+            'Unexpected deletion timestamp got {}'.format(
```
How is it `Unexpected deletion timestamp` when the point of the test is that you expect one?

Can be rewritten as:

```python
    def _check_volume_deletion_timestamp():
        volume = _get_volume(k8s_custom_client, name)
        assert volume is not None, 'Volume {} not found'.format(name)
        assert volume['metadata']['deletionTimestamp'] is not None,\
            'Volume {} is not marked for deletion'.format(name)
```

slaperche-scality commented on tests/post/features/volume.feature:
```diff
@@ -115,3 +115,88 @@ Feature: Volume management
+                operator: "Exists"
+                effect: "NoSchedule"
+        Then a pod 'pod-test-volume-e2e-2' exist with file '/mnt/toto.log'
+        # Enter the pod and check that the file exist
```
Is this implemented or not?
If so, remove the comment.

slaperche-scality commented on tests/post/steps/test_volume.py:
```diff
@@ -153,6 +157,49 @@ def delete_pvc(host, name, k8s_client):
+
+@when(parsers.parse("I delete the pod '{name}'"))
+def delete_namespaced_pod(host, name, k8s_client):
+    # ensure pod is available before deleting
```
No.

Pods are created with a `given`, the `given` should wait for availability. Not the delete.

slaperche-scality replied:
And for the wait on a pod, you should reuse `kube_utils.wait_for_pod`.

Ebaneck replied:
@slaperche-scality, i think i implemented the wait here because in the complete E2E test workflow, when we create the pod, it cannot be in available state because it has a pvc which is not yet fulfilled since the volume is only created in the `when`

Ebaneck replied:
Thinking we should create the volume in the `given` to furfill this check

BTW, there is nothing special about the volume definition.

Something like: 
```
    Scenario: Complete E2E volume workflow
        Given the following PersistentVolumeClaim 'e2e-pvc' exist:
         .....
        And a Volume 'volume9' exist```

slaperche-scality commented on tests/post/steps/test_volume.py:
```diff
@@ -94,6 +99,22 @@ def _check_status(status='Available'):
+    k8s_client.create_namespaced_persistent_volume_claim(
+        namespace='monitoring',
+        body=yaml.safe_load(body)
+    )
```
It's a `given` you should follow the same pattern as form Volume:

```
if exist(…):
    return
create(…)
wait for exist/available (depending on the object)
```

slaperche-scality commented on tests/post/steps/test_volume.py:
```diff
@@ -94,6 +99,22 @@ def _check_status(status='Available'):
+    k8s_client.create_namespaced_pod(
+        namespace='monitoring',
+        body=yaml.safe_load(body)
+    )
```
ditto

```
if exist(…):
    return
create(…)d
wait for exist/available (depending on the object)
```

Also, keep an empty line before closing `# }}}`.

slaperche-scality commented on tests/post/steps/test_volume.py:
```diff
@@ -303,6 +350,45 @@ def _check_deletion_timestamp():
+        _check_pod_status, times=30, wait=2,
+        name='checking status of Pod {}'.format(name)
+    )
+    with host.sudo():
```
I don't think you need to have a retry for the pod availability in a retry.

Just have a retry around the code that check the for the file.
Also, we may want to check file using the API instead of `kubectl` + we also want to check the content of the file.

Ebaneck replied:
Not sure, i can think of a better way to check the file using the API and it's content.

Can you enlighten me here??

slaperche-scality replied:
Take a look in `tests/install/steps/test_expansion.py`.

We're running an `etcdctl member list` command from inside the `etcd` container and get the output.

```python
etcd_member_list = etcdctl(k8s_client, ['member', 'list'], ssh_config)
[…]
def etcdctl(k8s_client, command, ssh_config):
    """Run an etcdctl command inside the etcd container."""
    name = 'etcd-{}'.format(
        utils.resolve_hostname('bootstrap', ssh_config)
    )
    etcd_command = [
        'etcdctl',
        '--endpoints', 'https://localhost:2379',
        '--ca-file', '/etc/kubernetes/pki/etcd/ca.crt',
        '--key-file', '/etc/kubernetes/pki/etcd/server.key',
        '--cert-file', '/etc/kubernetes/pki/etcd/server.crt',
    ] + command
    output = k8s.stream.stream(
        k8s_client.connect_get_namespaced_pod_exec,
        name=name, namespace='kube-system',
        command=etcd_command,
        stderr=True, stdin=False, stdout=True, tty=False
    )
    return output
```

Ebaneck commented on tests/post/features/volume.feature:
```diff
@@ -8,3 +8,194 @@ Feature: Volume management
+                  image:  busybox
+                  command:
+                  - sh
+                  args: ["-c", 'echo "toto" >> /mnt/toto.log && sleep 60']
```
Might be useful here to add a longer sleep interval since the container might exit after the sleep duration.
Also we have to account for the following
- pod creation time
- time for pod to become available(PV and PVC bound time)

I have noticed situations in which the check for pod availability fails.
And by the way, extending the sleep duration to something like 120s is not harmful since we delete this container thereafter.

slaperche-scality committed:
test/volume: add a fixture to cleanup volumes

To avoid unwanted side-effect and unexpected interactions between test
scenarios, we don't want volumes created into one scenario to be still
alive when the next one is executed.

To see why it's an issue, consider the latest test we introduced, where
we create PVC and expect it to claim the volume we just created. If old
volumes still lives around, the PVC may claim one those instead of the
one we just created => the test will now randomly fails and that's not
good.

Thus, we introduce a fixture which will cleanup any dangling volumes at
the end of each scenario.

Refs: #1417
Signed-off-by: Sylvain Laperche <sylvain.laperche@scality.com>

slaperche-scality committed:
test/volume: test volume accross Pod restart

If a Pod A mount a volume and writes data into it, another Pod B
mounting that same volume should see the data written by the previous
Pod.

Closes: #1417
Signed-off-by: Sylvain Laperche <sylvain.laperche@scality.com>

slaperche-scality committed:
test/volume: add pod & pvc cleanup

Some tests may leave dangling pods behind, and if those pods are using a
volume, they will prevent the cleanup of the volume.

Ditto for the PersistentVolumeClaim.

Thus, we first cleanup the pods, then the claims and finally we cleanup
the volumes.

Note that we renamed the fixture as it's no longer only about Volumes.

Refs: #1417
Signed-off-by: Sylvain Laperche <sylvain.laperche@scality.com>

slaperche-scality committed:
test: rename wait_for_pod

Wait for pod doesn't actually wait…
The name is misleading and must be changed.

https://www.luc-damas.fr/humeurs/nom-de-variable-explicite/

Signed-off-by: Sylvain Laperche <sylvain.laperche@scality.com>

slaperche-scality committed:
test/volume: code refactoring

Use classes to hide the subtle difference between each objects API in
order to be able to factorize common behavior (delete all, wait for
deletion, create from YAML, …)

Refs: #1417
Signed-off-by: Sylvain Laperche <sylvain.laperche@scality.com>

slaperche-scality committed:
test/volume: resolve nodeName for the Volume

Hostname need to be resolved since they are dynamic in the multinode
setup.

Refs: #1417
Signed-off-by: Sylvain Laperche <sylvain.laperche@scality.com>

bert-e commented:
# Waiting for approval



The following approvals are needed before I can proceed with the merge:


* the author


* one peer





Peer approvals *must* include at least 1 approval from the following list:

* @teddyandrieux

* @monpote

* @jbertran

* @sayf-eddine-scality

* @nicolast

* @cmonfort

* @slaperche-scality

* @thomasdanan

* @zempashi

* @lucieleonard

* @gdemonet

* @carlito-scality

* @nootal

* @alexandre-allard-scality

Ebaneck commented:
/approve

bert-e commented:
# In the queue



The changeset has received all authorizations and has been added to the
relevant queue(s). The queue(s) will be merged in the target development
branch(es) as soon as builds have passed.

The changeset will be merged in:


* :heavy_check_mark: `development/2.4`



The following branches will **NOT** be impacted:

* `development/1.0`
* `development/1.1`
* `development/1.2`
* `development/1.3`
* `development/2.0`
* `development/2.1`
* `development/2.2`
* `development/2.3`



There is no action required on your side. You will be notified here once
the changeset has been merged. In the unlikely event that the changeset
fails permanently on the queue, a member of the admin team will
contact you to help resolve the matter.

*IMPORTANT*

Please do not attempt to modify this pull request.

* Any commit you add on the source branch will trigger a new cycle after the
  current queue is merged.
* Any commit you add on one of the integration branches will be **lost**.

If you need this pull request to be removed from the queue, please contact a
member of the admin team now.




*The following options are set:* **approve**

bert-e commented:
# 


I have successfully merged the changeset of this pull request
into targetted development branches:


* :heavy_check_mark: `development/2.4`



The following branches have **NOT** changed:

* `development/1.0`
* `development/1.1`
* `development/1.2`
* `development/1.3`
* `development/2.0`
* `development/2.1`
* `development/2.2`
* `development/2.3`



Please check the status of the associated issue None.

Goodbye ebaneck.
